### PR TITLE
Implement state class for robot state management

### PR DESCRIPTION
**Pull request by zoecarver**
This patch adds a `State` class. This is how we will handle what our robot is supposed to be doing. We can update the state in the `update` method. It can be updated based on time, mode, buttons, etc. Then, in our periodic methods, we'll read the state and drive the robot accordingly.

### CODE REVIEW

File: state.py [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+			raise AttributeError
+		self.state = new_state
+	
+	""" 
+	Updates state to reflect the current state of the robot. This method 
+	should be called in every periodic method on the robot.
+	"""
+	def update(self):
+		""" kTeleop = 3; for now, just drive strait in teleop. """
+		if self.parent().Mode == 3:
```
Comment (zoecarver) [inline state.py:77]: I'll make `parent()` a property so this is a little cleaner and we can have proper errors once `parent` goes out of memory.

File: robot.py [hunk 2]
```diff
@@ -45,31 +47,39 @@ def robotInit(self):
...
         self.timer.reset()
         self.timer.start()
 
     def autonomousPeriodic(self):
         """Called every 20ms in autonomous mode."""
-        
-        # Drive forward for two seconds.
-        if self.timer.get() < 2:
-          self.myRobot.tankDrive(1, 1)
+        self._state.update()
```
Comment (julianallchin) [inline robot.py:67]: Might want to change the `state` class to `states` or something to avoid confusion with `_` characters.

File: robot.py [hunk 3]
```diff
@@ -79,13 +89,22 @@ def teleopPeriodic(self):
...
 
         # Tests the button debouncer
-        self.sd.putNumber((int) self.button.get())
+        self.sd.putNumber(self.button.get())
         
         # Tests ColorSensor output
         self.sd.putNumber(self.colorSensor.getGreen())
 
         # Tests the LimeLight's latency
         self.sd.putNumber(self.limelight.getLatency())
```
Comment (julianallchin) [inline robot.py]: `self.sd.putNumber()` requires a label before the value. i.e. `self.sd.putNumber("Button", self.button.get())`

File: state.py [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+""" Constants that represent various robot states """
+DRIVE_FORWARD_TWO_SEC = "DRIVE_FORWARD_TWO_SEC"
+TANK_DRIVE_NORMAL = "TANK_DRIVE_NORMAL"
+FREEZE = "FREEZE"
+
+"""
+\class StateOpts is a simple utility that acts as an enum wrapper for all 
+our states.
+"""
+class StateOpts(set):
```
Comment (julianallchin) [inline state.py:19]: I like this!

File: state.py [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+		It's really important that we have a weakref here. If the ref count
+		is incremented here, then we will create a circle reference and the 
+		garbage collector will never be able to delete either object creating
+		a leak.
+		"""
+		self.parent = weakref.ref(parent)
+		
+		self.opts = StateOpts()
+		
+		if default_state not in self.opts:
```
Comment (julianallchin) [inline state.py:62]: I believe we have been doing camelcasing. If you think that snake will be better we can do that

File: robot.py (hunk 3 above)
Comment (zoecarver) [reply robot.py]: Fixed on master

**Pull request merged**